gateway-discovery-cm/src/main/java/org/apache/knox/gateway/topology/discovery/cm/ClouderaManagerClusterConfigurationMonitor.java

```diff
@@ -0,0 +1,940 @@
+                                               lastTimestamp);
+    for (ApiEvent event : events) {
+      restartEvents.add(new RestartEvent(event));
+    }
+
+    return restartEvents;
+  }
+
+  private void persist(final Properties props, final File dest) {
+    try (OutputStream out = Files.newOutputStream(dest.toPath())) {
+      props.store(out, PERSISTED_FILE_COMMENT);
+      out.flush();
+    } catch (Exception e) {
+      log.failedToPersistClusterMonitorData(getType(), dest.getAbsolutePath(), e);
+    }
+  }
+
+  private File getPersistenceDir() {
+    File persistenceDir = null;
+
+    File dataDir = new File(gatewayConfig.getGatewayDataDir());
+    if (dataDir.exists()) {
+      File clustersDir = new File(dataDir, CLUSTERS_DATA_DIR_NAME);
+      if (!clustersDir.exists()) {
+        if (!clustersDir.mkdirs()) {
+          log.failedToCreatePersistenceDirectory(clustersDir.getAbsolutePath());
+        }
+      }
+      persistenceDir = clustersDir;
+    }
+
+    return persistenceDir;
+  }
+
+  /**
+   * Notify registered change listeners.
+   *
+   * @param source      The address of the ClouderaManager instance from which the cluster details were determined.
+   * @param clusterName The name of the cluster whose configuration details have changed.
+   */
+  private void notifyChangeListeners(final String source, final String clusterName) {
+    for (ConfigurationChangeListener listener : changeListeners) {
+      listener.onConfigurationChange(source, clusterName);
+    }
+  }
+
+  private void persistDiscoveryConfiguration(final String clusterName, final ServiceDiscoveryConfig sdc) {
+    File persistenceDir = getPersistenceDir();
+    if (persistenceDir != null) {
+
+      Properties props = new Properties();
+      props.setProperty(PROP_CLUSTER_NAME, clusterName);
+      props.setProperty(PROP_CLUSTER_SOURCE, sdc.getAddress());
+
+      String username = sdc.getUser();
+      if (username != null) {
+        props.setProperty(PROP_CLUSTER_USER, username);
+      }
+      String pwdAlias = sdc.getPasswordAlias();
+      if (pwdAlias != null) {
+        props.setProperty(PROP_CLUSTER_ALIAS, pwdAlias);
+      }
+
+      persist(props, getDiscoveryConfigPersistenceFile(sdc.getAddress(), clusterName));
+    }
+  }
+
+  private void persistServiceConfiguration(final String                                      discoveryAddress,
+                                           final String                                      cluster,
+                                           final Map<String, ServiceConfigurationProperties> configs) {
+    File persistenceDir = getPersistenceDir();
+    if (persistenceDir != null) {
+      File persistenceFile = getServiceConfigsPersistenceFile(discoveryAddress, cluster);
+      try ( ObjectOutputStream out = new ObjectOutputStream(Files.newOutputStream(persistenceFile.toPath())) ) {
+        out.writeObject(cluster);
+        out.writeObject(discoveryAddress);
+        out.writeObject(configs);
```

Comment:
This is same as readObject, should use JSON or something to avoid serializing Java classes directly.